fw/http2.c

```diff
@@ -489,6 +489,38 @@ tfw_h2_stream_xmit_prepare_resp(TfwStream *stream)
 	if (test_bit(TFW_HTTP_B_CHUNKED, resp->flags))
 		r = tfw_http_msg_cutoff_body_chunks(resp);
 
+	if (resp->trailers_len > 0) {
+		// MUST _NOT_ use resp pool to encode trailer headers
+		// because it assumes no headers after body in many code.
+		// Instead, encode trailers header in new skbs
+		// and append them to the resp->msg.skb_head.
+		struct sk_buff *skb_head, *skb, *nskb_head, *nskb;
+		unsigned long acc;
+
+		skb_head = resp->msg.skb_head;
+		skb = resp->mit.iter.skb;
+		resp->msg.skb_head = resp->mit.iter.skb = NULL;
+		acc = resp->mit.acc_len;
+
+		r = tfw_h2_hpack_encode_trailer_headers(resp);
```

Comment:
I don't agree with this solution. We should rework our standard functions to add trailers. Here we have extra loop over skbs and it looks bad for me. @krizhanovsky @const-t please look he and write your opinion.

Reply:
I don't see any problem here. I  implement such part of code
```
if (resp->trailers_len > 0) {
		{
			TfwHttpTransIter *mit = &resp->mit;
			TfwMsgIter *it = &mit->iter;
			
			it->frag = 4; // Setup it->frag or Iter->skb and frag which points to the end of the body. I set here 4, to check that test works fine.
 		}

		{
			unsigned long acc = resp->mit.acc_len;
			r = tfw_h2_hpack_encode_trailer_headers(resp);
			if (unlikely(r))
				goto finish;
			stream->xmit.t_len = resp->mit.acc_len - acc;
		}
```
And all works fine. So we don't need all previous code. We have two solutions:
1. Bad solution, not better performance that current solution, but less code. Use `tfw_http_iter_set_at
` or `tfw_msg_iter_move` to set at the end of the body.
2. Save pointer to the body and use it, may found body end during `tfw_http_msg_cutoff_body_chunks`

Reply:
If we really need to eliminate the "ugly" skb_head and mit manipulation, we can simply do `it->skb = NULL`, which triggers the new skb allocation to contain encoded trailers.
https://github.com/tempesta-tech/tempesta/blob/2a8d72d25aa638e58bdd2d0359d71c64f259ef9b/fw/http_msg.c#L1303

Reply:
We don't need to allocate extra skb if there is enough place in current skb. For short answers with trailers (which could be placed in one skb, you double the number of sending skbs, which is very inefficient.

Reply:
You can't use the response body's skb to hpack the trailer, because the skb is allocated from `pg_skb_alloc`, but `tfw_http_msg_expand_data` used by hpack assumes that you can use a whole page with one frag, which is conflicted to each other, and may cause the trailer to be truncated. Your successful case of reusing the frag in the test was nothing but a lucky case where the frag lounge just fit the trailer size. If you try to rerun the test with `-T 1`, you will see an error (frag=1, but the frag cannot actually accept the trailer). So all in all, a new skb seems to be the only feasible way. Note that we also can't use `resp->pool`, as mentioned repeatedly before, that way assumes that the response body follows the headers, i.e., that is a dedicated method for allocating normal headers and body, which does not work with trailers.

Reply:
It works! You should setup iterator->skb to the last body skb and iterator->frag to the last frag (which contains body!) of the last body skb. I write code below and it works! It is a dirty code to show that if you setup iterator all works fine. You should find a way to find the end of the body (skb and its last used fragment) and setup itertor!
```
{
			TfwHttpTransIter *mit = &resp->mit;
			TfwMsgIter *it = &mit->iter;
			
			struct sk_buff *skb = resp->msg.skb_head;
			do {
				if (skb->next == resp->msg.skb_head)
					break;
				skb = skb->next;
			} while(skb != resp->msg.skb_head);

			it->skb = skb;
			it->frag = skb_shinfo(it->skb)->nr_frags - 1;
		}

		{
			unsigned long acc = resp->mit.acc_len;
			TfwHttpTransIter *mit = &resp->mit;
			TfwMsgIter *it = &mit->iter;
			
			r = tfw_h2_hpack_encode_trailer_headers(resp);
			if (unlikely(r))
				goto finish;
			stream->xmit.t_len = resp->mit.acc_len - acc;
		}
```

Reply:
I also change increase body size in test and check it also for segmentation. If you setup iterator all works fine!

Reply:
I don't understand.

resp body skbs come from `pg_skb_alloc` used by the kernel from the backend socket receive path, once the frag is allocated, it cannot be extended in place, but a new chunk is allocated as a new frag.

```c
static int
__skb_fragment(struct sk_buff *skb_head, struct sk_buff *skb, char *pspt,
	       int len, TfwStr *it, int *fragn)
{
...
	/* See if the split starts in the page fragments data. */
	for (i = 0; i < si->nr_frags; ++i) {
		const skb_frag_t *frag = &si->frags[i];
		d_size = skb_frag_size(frag);
		offset = pspt - (char *)skb_frag_address(frag);

		/*
		 * @pspt can be the end of a frag, but it can also be a start of
		 * another fragment. Not necessary, that both frags are
		 * neighbours.
		 */
		if (offset >= 0 && offset < d_size) {
			len = max_t(long, len, offset - d_size);
			ret = __split_pgfrag(skb_head, skb, i, offset, len, it, fragn);
			goto done;
		}
		/*
		 * In case another fragment does not exist, remember the
		 * fragment number, after which you need to insert/delete data.
		 */
		if (offset == d_size)
			j = i;
	}
```
__skb_fragmentkb_fragment
```

In contrast, "tfw_http_msg_add_data" determines the fragment lounge, assuming it can use the entire page where the existing fragment resides.

```c
int
tfw_http_msg_add_data(TfwMsgIter *it, TfwHttpMsg *hm, TfwStr *field,
		      const TfwStr *data)
{
	const TfwStr *c, *end;

	BUG_ON(TFW_STR_DUP(data));
	if (WARN_ON_ONCE(it->frag >= skb_shinfo(it->skb)->nr_frags))
		return -E2BIG;

	TFW_STR_FOR_EACH_CHUNK(c, data, end) {
		char *p;
		unsigned int c_off = 0, c_size, f_room, n_copy;
this_chunk:
		c_size = c->len - c_off;
		if (it->frag >= 0) {
			unsigned int f_size;
			skb_frag_t *frag = &skb_shinfo(it->skb)->frags[it->frag];

			f_size = skb_frag_size(frag);
			f_room = PAGE_SIZE - frag->bv_offset - f_size;
			p = (char *)skb_frag_address(frag) + f_size;
			n_copy = min(c_size, f_room);
			skb_frag_size_add(frag, n_copy);
			ss_skb_adjust_data_len(it->skb, n_copy);
		} else {
			f_room = skb_tailroom(it->skb);
			n_copy = min(c_size, f_room);
			p = skb_put(it->skb, n_copy);
		}
```

We can see that the latter frag extend will overwrite the unallocated area of ​​the previous frag extend. Isn't it? So is it really safe to reuse skb between two different frag extend strategies?

Reply:
And also I check that you can use pool. You just should force allocate new frag in `tfw_http_msg_alloc_from_pool` and setup iterator.

```
static int
tfw_http_msg_alloc_from_pool(TfwHttpTransIter *mit, TfwPool* pool, size_t size, bool forse)
{
	if (!forse) {
		addr = tfw_pool_alloc_not_align_np(pool, size, &np);
	}
	else {
		addr = __tfw_pool_alloc_page(pool, size, /* align */ true);
		np = true;
	}
```

Also little rework of `__tfw_http_msg_expand_from_pool` is necessary in `if (unlikely(skb_room == 0 || nr_frags == MAX_SKB_FRAGS))` part of code.

Reply:
Here is a branch with example of code https://github.com/tempesta-tech/tempesta/tree/MekhanikEvgenii/fix-1902-dirty

Reply:
**As I raised my doubt above, the response has three different frag allocation strategies:**
`headers (per-cpu page cache) --> body (pg_skb_alloc) --> trailers (kernel pages)`.

**So no matter use pool or not, and no matter reuse the last skb of the body to hpack trailers or not, we cannot reuse the last frag of the body.**

I make a small change to consider both skb reuse and frag distinguish, please check this [commit](https://github.com/tempesta-tech/tempesta/pull/2173/commits/3d06edbbd2a58f319e06e8851ad8ea2e08c4e330). Hopefully it satisfy your need.

---

_P.S. no need to rework the pool functions to adapt trailers, because it has the assumption of the header-body order, if we rework it, the functions will be too much fan in, which seems not a good coding practice. And no offense, but your example code requires much more work to get it to work, where I also tried before._

Reply:
I don't see problems in different allocation strategies. @const-t please pay attention on this place. Moreover in my example of usage pool  we don't reuse the last frag of the body, but allocate new one. And in my example we don't need a big part of code in `tfw_hpack_hdr_expand`. Much more work is not a problem as I think.

Reply:
[done](https://github.com/tempesta-tech/tempesta/pull/2173/commits/bb1589a70478b69aff2a7a455c72da865a0ec4bd).

Reply:
Can you re-think if it's ok to reuse the skb/frag? They use different allocations, if we mix them, we may override something. Please check my comments above. Anyway, if you're all 100% sure, I'll not insist, then please go ahead with one of your solutions.
`headers (pool, i.e. per-cpu page cache) --> body (pg_skb_alloc) --> trailers (kernel pages)`

Reply:
I've seen this patch. However, from my point of view, would be better to do it another way, without changing `tfw_http_msg_expand_data()` and without code duplication in `tfw_hpack_hdr_expand()`. We must using pool and use `tfw_hpack_transform()`. There are few reasons why it's better:
1. We don't change common purpose function `tfw_http_msg_expand_data()` only for trailers case.
2. We don't add [this code](https://github.com/tempesta-tech/tempesta/blob/bfe4c9cab5075d8ced311571299e308cf780b660/fw/hpack.c#L3492).
3. We could use dynamic table, because now we can't do it.
4. Preparation of pool will be located in same place where we do other work with trailers.

Reply:
Fine with me. Submit your commit directly or provide suggested changes please, thank you.

Reply:
I fix this part of code. @const-t can you look on it?